File: pc-frontend/src/features/assets/eskAssetApi.ts
import { api } from '../../api/client'

export interface EskAssetIdentity {
  asset_id: 'esk'
  symbol: 'ESK'
  name: string
  decimals: 6
  issuance_mode: 'paper_recorded'
  chain_status: 'not_deployed'
  contract_address: null
}

export interface EskAssetBalance {
  total: string
  available: string
  reserved_for_sellback: string
  reserved_for_quant: string
  reserved_total: string
  total_base_units: string
  available_base_units: string
  sellback_reserved_base_units: string
  quant_reserved_base_units: string
  reserved_base_units: string
  revision: number
  updated_at: string | null
}

export interface EskAssetSnapshot {
  schema: 'yilong.esk.asset_account.v2'
  mode: 'disabled' | 'paper' | 'invalid'
  enabled: boolean
  simulated: true
  funds_moved: false
  asset: EskAssetIdentity
  balance: EskAssetBalance
  sellback: {
    application_only: true
    request_enabled: boolean
    settlement_enabled: false
    pricing_status: 'not_defined'
  }
  status_message: string
}

export interface EskQuantAllocationRequest {
  request_id: string
  amount: string
  amount_base_units: string
  risk_disclosure_revision: 'esk-quant-paper-allocation-v2'
  status: 'submitted' | 'canceled' | 'accepted' | 'released'
  revision: number
  submitted_at: string
  updated_at: string
  simulated: true
  funds_moved: false
  position_created: false
  allocation_binding_created: boolean
  binding_id?: string
  quant_binding_revision?: number
  occurred_at_unix?: number
  replayed: boolean
}

interface EskQuantAllocationList {
  schema: 'yilong.esk.quant_allocation_request_list.v2'
  simulated: true
  funds_moved: false
  position_created: false
  requests: EskQuantAllocationRequest[]
}

export interface EskSellbackRequest {
  request_id: string
  amount: string
  amount_base_units: string
  status: 'submitted' | 'canceled'
  revision: number
  submitted_at: string
  updated_at: string
  simulated: true
  funds_moved: false
  replayed: boolean
}

interface EskSellbackList {
  schema: 'yilong.esk.sellback_request_list.v1'
  simulated: true
  funds_moved: false
  requests: EskSellbackRequest[]
}

export type EskExchangeDirection = 'usdt_to_esk' | 'esk_to_usdt'

export interface EskExchangeAccount {
  schema: 'yilong.esk.paper_exchange_account.v1'
  mode: 'disabled' | 'paper' | 'invalid'
  enabled: boolean
  simulated: true
  funds_moved: false
  on_chain_settlement: false
  trading_mode: 'paper'
  balances: {
    esk: { total: string; available: string; revision: number; updated_at: string | null }
    usdt: { total: string; available: string; revision: number; updated_at: string | null }
  }
  pricing: null | {
    usdt_per_esk: string
    fee_bps: number
    fee_percent: string
    config_revision: string
    quote_ttl_seconds: 60
  }
  status_message: string
}

export interface EskExchangeQuote {
  schema: 'yilong.esk.paper_exchange_quote.v1'
  quote_id: string
  direction: EskExchangeDirection
  input_asset: 'ESK' | 'USDT'
  output_asset: 'ESK' | 'USDT'
  input_amount: string
  gross_output_amount: string
  fee_asset: 'ESK' | 'USDT'
  fee_amount: string
  net_output_amount: string
  usdt_per_esk: string
  fee_bps: number
  created_at: string
  expires_at: string
  simulated: true
  funds_moved: false
  on_chain_settlement: false
  trading_mode: 'paper'
}

export interface EskExchangeExecution {
  schema: 'yilong.esk.paper_exchange_execution.v1'
  execution_id: string
  executed_at: string
  replayed: boolean
  quote: EskExchangeQuote
  simulated: true
  funds_moved: false
  on_chain_settlement: false
  trading_mode: 'paper'
}

interface EskExchangeExecutionList {
  schema: 'yilong.esk.paper_exchange_execution_list.v1'
  simulated: true
  funds_moved: false
  on_chain_settlement: false
  trading_mode: 'paper'
  executions: EskExchangeExecution[]
}

export const eskAssetApi = {
  account: () => api.get<EskAssetSnapshot>('/api/me/assets/esk'),
  sellbackRequests: () => api.get<EskSellbackList>('/api/me/assets/esk/sellback-requests?limit=20'),
  quantAllocationRequests: () => api.get<EskQuantAllocationList>('/api/me/assets/esk/quant-allocation-requests?limit=20'),
  exchangeAccount: () => api.get<EskExchangeAccount>('/api/me/assets/esk/exchange-account'),
  exchangeHistory: () => api.get<EskExchangeExecutionList>('/api/me/assets/esk/exchanges?limit=20'),
  createExchangeQuote: (direction: EskExchangeDirection, inputAmount: string) =>
    api.post<EskExchangeQuote>('/api/me/assets/esk/exchange-quotes', {
      direction,
      input_amount: inputAmount,
    }),
  executeExchange: (quoteId: string, idempotencyKey: string) =>
    api.post<EskExchangeExecution>('/api/me/assets/esk/exchanges', {
      quote_id: quoteId,
      idempotency_key: idempotencyKey,
      confirmation: 'CONFIRM PAPER ESK USDT EXCHANGE',
    }),
  createSellback: (amount: string, idempotencyKey: string) =>
    api.post<EskSellbackRequest>('/api/me/assets/esk/sellback-requests', {
      amount,
      idempotency_key: idempotencyKey,
    }),
  cancelSellback: (requestId: string) =>
    api.post<EskSellbackRequest>(`/api/me/assets/esk/sellback-requests/${encodeURIComponent(requestId)}/cancel`, {
      confirmation: 'CANCEL ESK SELLBACK REQUEST',
    }),
  createQuantAllocation: (amount: string, idempotencyKey: string) =>
    api.post<EskQuantAllocationRequest>('/api/me/assets/esk/quant-allocation-requests', {
      amount,
      idempotency_key: idempotencyKey,
      risk_disclosure_revision: 'esk-quant-paper-allocation-v2',
      confirmation: 'REQUEST PAPER ESK QUANT ALLOCATION',
    }),
  cancelQuantAllocation: (requestId: string) =>
    api.post<EskQuantAllocationRequest>(`/api/me/assets/esk/quant-allocation-requests/${encodeURIComponent(requestId)}/cancel`, {
      confirmation: 'CANCEL PAPER ESK QUANT ALLOCATION',
    }),
  applyQuantAllocationReceipt: (receiptToken: string) =>
    api.post<EskQuantAllocationRequest>('/api/me/assets/esk/quant-allocation-receipts', {
      receipt_token: receiptToken,
    }),
}

export const ESK_PREVIEW_SNAPSHOT: EskAssetSnapshot = {
  schema: 'yilong.esk.asset_account.v2',
  mode: 'paper',
  enabled: true,
  simulated: true,
  funds_moved: false,
  asset: {
    asset_id: 'esk',
    symbol: 'ESK',
    name: '一龙 ESK',
    decimals: 6,
    issuance_mode: 'paper_recorded',
    chain_status: 'not_deployed',
    contract_address: null,
  },
  balance: {
    total: '1280.000000',
    available: '780.000000',
    reserved_for_sellback: '200.000000',
    reserved_for_quant: '300.000000',
    reserved_total: '500.000000',
    total_base_units: '1280000000',
    available_base_units: '780000000',
    sellback_reserved_base_units: '200000000',
    quant_reserved_base_units: '300000000',
    reserved_base_units: '500000000',
    revision: 4,
    updated_at: '2026-09-02T04:30:00Z',
  },
  sellback: {
    application_only: true,
    request_enabled: true,
    settlement_enabled: false,
    pricing_status: 'not_defined',
  },
  status_message: 'Paper 测试登记，尚未上链；可用余额已扣除卖回和量化申请占用，两类申请都不代表成交、付款、入金或收益。',
}

export const ESK_PREVIEW_REQUESTS: EskSellbackRequest[] = [{
  request_id: 'eskr_preview_001',
  amount: '200.000000',
  amount_base_units: '200000000',
  status: 'submitted',
  revision: 1,
  submitted_at: '2026-09-02T04:35:00Z',
  updated_at: '2026-09-02T04:35:00Z',
  simulated: true,
  funds_moved: false,
  replayed: false,
}]

export const ESK_PREVIEW_QUANT_REQUESTS: EskQuantAllocationRequest[] = [{
  request_id: 'eskq_preview_001',
  amount: '300.000000',
  amount_base_units: '300000000',
  risk_disclosure_revision: 'esk-quant-paper-allocation-v2',
  status: 'submitted',
  revision: 1,
  submitted_at: '2026-09-02T04:40:00Z',
  updated_at: '2026-09-02T04:40:00Z',
  simulated: true,
  funds_moved: false,
  position_created: false,
  allocation_binding_created: false,
  replayed: false,
}]

File: android/app/src/main/kotlin/com/elon/app/esk/EskAssetApi.kt
package com.elon.app.esk

import android.content.Context
import com.elon.app.AuthManager
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

internal data class EskAssetSnapshot(
    val mode: String,
    val enabled: Boolean,
    val total: String,
    val available: String,
    val reservedForSellback: String,
    val reservedForQuant: String,
    val reservedTotal: String,
    val balanceRevision: Long,
    val updatedAt: String?,
    val sellbackRequestEnabled: Boolean,
    val chainStatus: String,
    val statusMessage: String,
)

internal data class EskSellbackRequest(
    val requestId: String,
    val amount: String,
    val status: String,
    val submittedAt: String,
)

internal data class EskExchangeAccount(
    val enabled: Boolean,
    val eskAvailable: String,
    val usdtAvailable: String,
    val usdtPerEsk: String?,
    val feePercent: String?,
    val statusMessage: String,
)

internal data class EskExchangeQuote(
    val quoteId: String,
    val direction: String,
    val inputAsset: String,
    val outputAsset: String,
    val inputAmount: String,
    val grossOutputAmount: String,
    val feeAmount: String,
    val netOutputAmount: String,
    val expiresAt: String,
)

internal data class EskExchangeExecution(
    val executionId: String,
    val quote: EskExchangeQuote,
)

internal class EskAssetApi(
    private val context: Context,
    private val http: OkHttpClient,
    private val serverUrl: String,
) {
    fun account(): EskAssetSnapshot {
        val root = execute(authenticated(Request.Builder().url(url("/api/me/assets/esk")).get()))
        val asset = root.optJSONObject("asset") ?: error("ESK 资产身份缺失")
        val balance = root.optJSONObject("balance") ?: error("ESK 余额缺失")
        val sellback = root.optJSONObject("sellback") ?: error("ESK 卖回策略缺失")
        require(root.optString("schema") == "yilong.esk.asset_account.v2") {
            "ESK 资产协议版本不匹配"
        }
        require(asset.optString("symbol") == "ESK") { "ESK 资产标识不匹配" }
        require(asset.optString("chain_status") == "not_deployed") { "ESK 上链状态不匹配" }
        require(root.optBoolean("simulated") && !root.optBoolean("funds_moved", true)) {
            "ESK Paper 安全标识不匹配"
        }
        return EskAssetSnapshot(
            mode = root.optString("mode", "invalid"),
            enabled = root.optBoolean("enabled"),
            total = exactAmount(balance, "total"),
            available = exactAmount(balance, "available"),
            reservedForSellback = exactAmount(balance, "reserved_for_sellback"),
            reservedForQuant = exactAmount(balance, "reserved_for_quant"),
            reservedTotal = exactAmount(balance, "reserved_total"),
            balanceRevision = balance.optLong("revision").takeIf { balance.has("revision") && it >= 0 }
                ?: error("ESK 余额修订无效"),
            updatedAt = balance.optString("updated_at").ifBlank { null },
            sellbackRequestEnabled = sellback.optBoolean("request_enabled"),
            chainStatus = asset.optString("chain_status", "unknown"),
            statusMessage = root.optString("status_message", "ESK 资产状态未知"),
        )
    }

    fun requests(): List<EskSellbackRequest> {
        val root = execute(authenticated(Request.Builder().url(url("/api/me/assets/esk/sellback-requests?limit=20")).get()))
        require(root.optBoolean("simulated") && !root.optBoolean("funds_moved", true)) {
            "ESK 卖回申请安全标识不匹配"
        }
        val values = root.optJSONArray("requests") ?: return emptyList()
        return (0 until values.length()).mapNotNull { index ->
            values.optJSONObject(index)?.let(::parseRequest)
        }
    }

    fun exchangeAccount(): EskExchangeAccount {
        val root = execute(authenticated(Request.Builder().url(url("/api/me/assets/esk/exchange-account")).get()))
        requirePaperExchangeEnvelope(root, "yilong.esk.paper_exchange_account.v1")
        val balances = root.optJSONObject("balances") ?: error("Paper 兑换余额缺失")
        val esk = balances.optJSONObject("esk") ?: error("ESK Paper 兑换余额缺失")
        val usdt = balances.optJSONObject("usdt") ?: error("USDT Paper 兑换余额缺失")
        val pricing = root.optJSONObject("pricing")
        return EskExchangeAccount(
            enabled = root.optBoolean("enabled") && pricing != null,
            eskAvailable = exactAmount(esk, "available"),
            usdtAvailable = exactAmount(usdt, "available"),
            usdtPerEsk = pricing?.let { exactAmount(it, "usdt_per_esk") },
            feePercent = pricing?.optString("fee_percent")?.takeIf { it.matches(FEE_PERCENT) },
            statusMessage = root.optString("status_message", "Paper 兑换状态未知"),
        )
    }

    fun createExchangeQuote(direction: String, inputAmount: String): EskExchangeQuote {
        require(direction == "usdt_to_esk" || direction == "esk_to_usdt") { "兑换方向无效" }
        val body = JSONObject().put("direction", direction).put("input_amount", inputAmount)
        val root = execute(authenticated(
            Request.Builder().url(url("/api/me/assets/esk/exchange-quotes"))
                .post(body.toString().toRequestBody(JSON)),
        ))
        return parseExchangeQuote(root)
    }

    fun executeExchange(quoteId: String, idempotencyKey: String): EskExchangeExecution {
        val body = JSONObject()
            .put("quote_id", quoteId)
            .put("idempotency_key", idempotencyKey)
            .put("confirmation", "CONFIRM PAPER ESK USDT EXCHANGE")
        val root = execute(authenticated(
            Request.Builder().url(url("/api/me/assets/esk/exchanges"))
                .post(body.toString().toRequestBody(JSON)),
        ))
        requirePaperExchangeEnvelope(root, "yilong.esk.paper_exchange_execution.v1")
        return EskExchangeExecution(
            executionId = root.optString("execution_id").ifBlank { error("Paper 兑换流水号缺失") },
            quote = parseExchangeQuote(root.optJSONObject("quote") ?: error("Paper 兑换回执缺少报价")),
        )
    }

    fun createSellback(amount: String, idempotencyKey: String): EskSellbackRequest {
        val body = JSONObject().put("amount", amount).put("idempotency_key", idempotencyKey)
        return parseRequest(execute(authenticated(
            Request.Builder().url(url("/api/me/assets/esk/sellback-requests"))
                .post(body.toString().toRequestBody(JSON)),
        )))
    }

    fun cancelSellback(requestId: String): EskSellbackRequest {
        val body = JSONObject().put("confirmation", "CANCEL ESK SELLBACK REQUEST")
        val encoded = android.net.Uri.encode(requestId)
        return parseRequest(execute(authenticated(
            Request.Builder().url(url("/api/me/assets/esk/sellback-requests/$encoded/cancel"))
                .post(body.toString().toRequestBody(JSON)),
        )))
    }

    private fun parseRequest(value: JSONObject): EskSellbackRequest {
        require(value.optBoolean("simulated") && !value.optBoolean("funds_moved", true)) {
            "ESK 卖回申请安全标识不匹配"
        }
        return EskSellbackRequest(
            requestId = value.optString("request_id"),
            amount = exactAmount(value, "amount"),
            status = value.optString("status"),
            submittedAt = value.optString("submitted_at"),
        )
    }

    private fun parseExchangeQuote(value: JSONObject): EskExchangeQuote {
        requirePaperExchangeEnvelope(value, "yilong.esk.paper_exchange_quote.v1")
        val inputAsset = assetSymbol(value, "input_asset")
        val outputAsset = assetSymbol(value, "output_asset")
        require(value.optString("fee_asset") == outputAsset) { "Paper 兑换手续费资产不匹配" }
        return EskExchangeQuote(
            quoteId = value.optString("quote_id").ifBlank { error("Paper 兑换报价 ID 缺失") },
            direction = value.optString("direction"),
            inputAsset = inputAsset,
            outputAsset = outputAsset,
            inputAmount = exactAmount(value, "input_amount"),
            grossOutputAmount = exactAmount(value, "gross_output_amount"),
            feeAmount = exactAmount(value, "fee_amount"),
            netOutputAmount = exactAmount(value, "net_output_amount"),
            expiresAt = value.optString("expires_at").ifBlank { error("Paper 兑换报价有效期缺失") },
        )
    }

    private fun requirePaperExchangeEnvelope(value: JSONObject, schema: String) {
        require(value.optString("schema") == schema) { "Paper 兑换协议版本不匹配" }
        require(value.optBoolean("simulated") && !value.optBoolean("funds_moved", true)) {
            "Paper 兑换资金安全标识不匹配"
        }
        require(!value.optBoolean("on_chain_settlement", true) && value.optString("trading_mode") == "paper") {
            "Paper 兑换链上安全标识不匹配"
        }
    }

    private fun assetSymbol(value: JSONObject, key: String): String =
        value.optString(key).takeIf { it == "ESK" || it == "USDT" }
            ?: error("Paper 兑换资产标识无效")

    private fun exactAmount(value: JSONObject, key: String): String =
        value.optString(key).takeIf { it.matches(EXACT_AMOUNT) }
            ?: error("ESK 金额格式无效")

    private fun authenticated(builder: Request.Builder): Request =
        AuthManager.applyAuth(context, builder).build()

    private fun execute(request: Request): JSONObject = http.newCall(request).execute().use { response ->
        val text = response.body?.string().orEmpty()
        val json = runCatching { JSONObject(text.ifBlank { "{}" }) }.getOrDefault(JSONObject())
        if (!response.isSuccessful) {
            error(json.optString("error").ifBlank { "请求失败 (${response.code})" }.take(500))
        }
        json
    }

    private fun url(path: String) = serverUrl.trimEnd('/') + path

    private companion object {
        val JSON = "application/json".toMediaType()
        val EXACT_AMOUNT = Regex("^\\d+\\.\\d{6}$")
        val FEE_PERCENT = Regex("^\\d+\\.\\d{2}%$")
    }
}
